speaker/layer2_controller.go

```diff
@@ -75,14 +80,14 @@ func usableNodes(eps *v1.Endpoints, mList *memberlist.Memberlist) []string {
 	return ret
 }
 
-func (c *layer2Controller) ShouldAnnounce(l log.Logger, name string, svc *v1.Service, eps *v1.Endpoints) string {
-	nodes := usableNodes(eps, c.mList)
-	// Sort the slice by the hash of node + service name. This
-	// produces an ordering of ready nodes that is unique to this
-	// service.
+func (c *layer2Controller) ShouldAnnounce(l log.Logger, svc *v1.Service, eps *v1.Endpoints, lbIP net.IP) string {
+	nodes := usableNodes(svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeLocal, eps, c.mList)
+	// Sort the slice by the hash of node + service IP. This
+	// produces an ordering of ready nodes that is unique to the IP
+	// so that all services sharing the same IP are annonced from the same node.
 	sort.Slice(nodes, func(i, j int) bool {
-		hi := sha256.Sum256([]byte(nodes[i] + "#" + name))
-		hj := sha256.Sum256([]byte(nodes[j] + "#" + name))
+		hi := sha256.Sum256([]byte(nodes[i] + "#" + lbIP.String()))
+		hj := sha256.Sum256([]byte(nodes[j] + "#" + lbIP.String()))
```

Comment:
We should probably add a release note to let people know that the algorithm changes and will result in many services having their IPs move.  This could cause some brief downtime.

I also imagine this sort of change should only come in a new `Y` release of `X.Y.Z`.

@rata thoughts?